edb/schema/pointers.py

```diff
@@ -1545,6 +1560,80 @@ def _cmd_tree_from_ast(
         return cmd
 
 
+class AlterPointerUpperCardinality(
+    referencing.ReferencedInheritingObjectCommand[Pointer_T],
+    inheriting.AlterInheritingObjectFragment[Pointer_T],
+    sd.AlterSpecialObjectField[Pointer_T],
+    PointerCommandOrFragment[Pointer_T],
+):
+    """Handler for the "cardinality" field changes."""
+
+    def _alter_begin(
+        self,
+        schema: s_schema.Schema,
+        context: sd.CommandContext,
+    ) -> s_schema.Schema:
+        orig_schema = schema
+        schema = super()._alter_begin(schema, context)
+        scls = self.scls
+
+        orig_card = scls.get_cardinality(orig_schema)
+        new_card = scls.get_cardinality(schema)
+
+        if orig_card == new_card:
+            # The actual value hasn't changed, nothing to do here.
+            return schema
+
+        # Cardinality alters of pointers used in expressions is prohibited.
+        # Eventually we may be able to relax this by revalidating the
+        # expressions and making sure their meaning and cardinality does not
+        # change.
+        vn = scls.get_verbosename(schema, with_parent=True)
+        schema = self._prohibit_if_expr_refs(
+            schema, context, action=f'alter the cardinality of {vn}')
+
+        if not context.canonical:
+            implicit_bases = scls.get_implicit_bases(schema)
+            non_altered_bases = []
+
+            for base in set(implicit_bases) - context.altered_targets:
+                base_card = base.get_cardinality(schema)
+                if base_card != new_card:
+                    non_altered_bases.append(base)
+
+            # This pointer is inherited from one or more ancestors that
+            # are not altered in the same op, and this is an error.
+            if non_altered_bases:
+                bases_str = ', '.join(
+                    b.get_verbosename(schema, with_parent=True)
+                    for b in non_altered_bases
+                )
+
+                vn = scls.get_verbosename(schema, with_parent=True)
+
+                raise errors.SchemaDefinitionError(
```

Comment:
This is new behavior, right? Add a test?

Reply:
+1